mesonbuild/build.py

```diff
@@ -373,15 +394,28 @@ class IncludeDirs(HoldableObject):
     # Interpreter has validated that all given directories
     # actually exist.
     extra_build_dirs: T.List[str] = field(default_factory=list)
+    use_type: IncludeDirsUse = field(default=IncludeDirsUse.src_and_build_relative)
+
+    # It's meaningless to explicitly specify 'build_relative' include dirs but with absolute paths.
+    def _validate_no_build_relative_abs_paths(self):
+        if self.use_type == IncludeDirsUse.build_relative:
+            for d in self.incdirs:
+                if os.path.isabs(d):
+                    raise MesonException(f'IncludeDirs, with \'build_relative\', using absolute path ({d}), is not expected. This is likely a user error.')
```

Comment:
This is the wrong place to do use input validation, this should all be happening in the Interpreter/module layer, and the build layer should assume what it's getting is valid already.

Reply:
We have instantiation of IncludeDirs in several places -
- modules/windows.py, `compile_resources(...)`
- modules/gnome.py, `generate_vapi(...)`
- interpreter/interpreterobjects.py, `private_dir_include_method(...)`
- build.py, `add_include_dirs(...)`
- interpereter/interpreter.py, `build_incdir_object(...)`

I think the last 2 of these are the main instances of user-script-supplied include dirs, which would most benefit from validating that the user isn't likely confused by using an absolute path with the 'build'-relative use_type.

Are you suggesting I move `_validate_no_build_relative_abs_paths(...)` outside of `IncludeDirs` to a free function that is then explicitly called prior to each of the instantiation scenarios above (or perhaps just the last one)?

And then are there any concerns with the possibility of missing this bit of user sanity checking if some future bit of interpreter functionality decides to construct more new `IncludeDirs` instances while negecting to perform this bit of validation?

I ask these because, with the current state of these changes, this validation is very difficult to avoid all current and possible future use of `IncludeDirs`, which seems a reasonable thing to do, so I'd like to better understand the actual problems with being done where it is as opposed to in one or more places elsewhere.

Reply:
> the build layer should assume what it's getting is valid already

> Are you suggesting [...] a free function that is then explicitly called prior to each of the instantiation scenarios above (or perhaps just the last one)?

No, the suggestion is that instantiation scenarios do not need any validation at all, because it's assumed to be valid.

There's a conflict of understanding here, in that:
- @dcbaker is assuming the point of validation is to check that users give valid inputs to meson.build, but the meson code itself contains no bugs, for example because it's tested using unittests
- you are assuming the point of validation is to check that future contributors to the meson codebase don't violate the invariants of the API.

My hot take is that doing sanity checking like this for the latter case is fundamentally incorrect and wrong, in all domains and all languages, not because of the reasons one might expect, but because it should be using an assert for code that is only meant to run in debug builds to verify that other coders working on the codebase didn't write incorrect code.

:P

Reply:
I see.  In which case, we end up with one assertion validation internal to IncludeDirs, similar to what's already there but an assert, in '__post_init__', instead of a MesonException.  Then another check, doing exactly the same validation logic, only ending up with a user-facing MesonException message, to be done only on the interpreter's processing of the user's dirs listed in their `include_directories(...)` input, right?

Now, because both uses want to perform exactly the same checks under slightly different circumstances, instead of duplicating the same checking logic in 2 difeerent places, it would seem reasonable to add a IncludeDirs method (maybe something called `validate_no_build_relative_abs_paths()` or `has_no_build_relative_abs_paths()`) that's used in both cases.  
This is very similar to what's already there, seems reasonable, and I'm fine with making this change.

Reply:
Sure. I mean, my main focus here wasn't to take a side and say that it either should or should not check in both places... I'm currently just focused on helping everyone to better understand what the proposal *is*. That being said, I don't personally mind an assert, assuming no other devs have a strong objection.

So in the interest of better understanding what the proposal is, I think it's very beneficial for code review *in general* to be able to look at a piece of code and say "this is an assert, clearly the author of the changeset intends for this to be testing an invariant, the violation of which constitutes a meson bug", and distinguish that from "this code appears to be testing user input".

Reviewers might then respond to the assert by saying "I don't think this assert is needed", or they might respond by saying "okay now I get it, that's fine to have there". But they won't respond by saying "user input can't get here why are you checking it again". :)

So I think always being clear about what's an invariant check and what's a user input validation check, is important.